sram: add example of mocking SRAMs to asap7/riscv32i

oharboe opened the pull request:
Mocked SRAMs and fake srams give very similar results, i.e. for design exploration mocking SRAMs can be more than adequate. The mocking of SRAMs could be refined to be more similar to the fake sram results.

With mocked SRAMs:

![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/2798822/7dde1cb3-27ef-4f54-82f5-77827e0ebcf0)

Compared to fake srams:

![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/2798822/06800dfe-6ae3-470f-99bf-707490f20197)

Also, the smallest positive slack timing paths are almost the same.

![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/2798822/b3b676d2-3b84-4d0d-9a4f-f87dd64931f8)

oharboe committed:
sram: add SRAM mocking example asap7/riscv32i-mock-sram

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

maliberty commented on flow/designs/asap7/riscv32i/config.mk:
With no CI I wonder if this will continue to work.  It might be better to convert the test case to use this if this is important.

oharboe replied:
Fixed. Force pushed to make it easier to review as I've changed the structure of the PR.

oharboe committed:
asap7: add generic constraints.sdc file to platforms folder

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

rovinski commented on flow/designs/asap7/mock-array/constraints.sdc:
```diff
@@ -1,32 +1,5 @@
-#  See platforms/asap7/lib/asap7sc7p5t_INVBUF_RVT_FF_nldm_220122.lib.gz, line 1223.
-#set_load -pin_load 2.731126 [all_outputs]
-set_load -pin_load 10 [all_outputs]
+source $env(PLATFORM_DIR)/constraints.sdc
```
Small nit but I think that `source` is not valid in SDC. It would only work if the SDC is read as a normal Tcl script.

oharboe replied:
The "source" is used to reduce reptition, but also to document the shared/generic aspect of this cosntraints.sdc that is only used for architectural exploration purposes.

It works with OpenROAD and ORFS, so good 'nuf to use around the house?

I would expect mocking/modeling to be EDA toolchain specific...

Also, floorplan will read in the script and then write out an .sdc file after which the the "source" goes away. I presume that OpenSTA will write out the .sdc file in some canonical form.

maliberty commented on flow/platforms/asap7/constraints.sdc:
```diff
@@ -0,0 +1,82 @@
+#  pin capacitance for this cell, which is 0.577042.
+#  See platforms/asap7/lib/asap7sc7p5t_INVBUF_RVT_FF_nldm_220122.lib.gz, line 1223.
+#set_load -pin_load 2.731126 [all_outputs]
+set_load -pin_load 10 [all_outputs]
```
This comment looks stale.  How was 10 arrived at?

oharboe replied:
I dont know.... @tspyrou Thoughts?

maliberty commented on flow/platforms/asap7/constraints.sdc:
```diff
@@ -0,0 +1,82 @@
+set all_register_outputs [get_pins -of_objects [all_registers] -filter {direction == output}]
+
+# Optimization targets. This 
+set max_delay 80
```
Why 80?

oharboe replied:
I want it to be on the order of the time it takes to traverse a small macro(io-io) and to go from a reg to io and io to reg. Better over, than underconstrained.

Thoughts?

@tspyrou ?

oharboe committed:
riscv32i-mock-sram/fakeram7_256x32: reduce congestion and sensitivity to initial conditions

very slightly changing values would cause global route to fail
because of high congestion. Loose congestion by placing pins
further apart

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
asap7/constraints.sdc: improve documentation

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
asap7/constraints.sdc: remove set_load and set_driving_cell

this constraints.sdc file doesn't have that level of sophistication,
it is only used for architectural exploration.

It is intended to be a minimal constraints.sdc file that
should be replaced with a more refined specific version
for real designs.

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
riscv32i-mock-sram/fakeram7_256x32: add CI

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
asap/constraints.sdc: delete out of scope details

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
test_helper: fix error with subfolder designs, such as riscv32i-mock-sram/fakeram7_256x32

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
riscv32i-mock-sram/fakeram7_256x32: update_ok

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>
